src/Distribution/Nixpkgs/Meta.hs

```diff
@@ -85,6 +92,79 @@ nixpkgsPlatformFromCabal (Platform arch os) = "\"" ++ nixArch ++ "-" ++ nixOs ++
+--   A 'Just' result thus only indicates that the format of the platform is
+--   sound — 'nixpkgsPlatformFromString' does /not/ check if the parsed platform
+--   actually exists.
+--
+--   __Note__: 'nixpkgsPlatformFromString' is /not/ the inverse operation for
+--   'NixpkgsPlatform'\'s 'Pretty' instance. It is not intended for parsing Nix
+--   expressions.
+--
+--   >>> nixpkgsPlatformFromString "x86_64-netbsd"
+--   Just (NixpkgsPlatformSingle (Platform X86_64 NetBSD))
+--   >>> nixpkgsPlatformFromString "platforms.riscv"
+--   Just (NixpkgsPlatformGroup (Identifier "riscv"))
+--   >>> nixpkgsPlatformFromString "garbage"
+--   Nothing
+nixpkgsPlatformFromString :: String -> Maybe NixpkgsPlatform
+nixpkgsPlatformFromString s = platformGroup <|> singlePlatform
+  where platformGroup = do
+          -- also accept "platform." as prefix to save some typing
+          name <- stripPrefix "lib.platforms." s <|> stripPrefix "platforms." s
+          Just $ NixpkgsPlatformGroup (ident # name)
+
+        singlePlatform = NixpkgsPlatformSingle <$> cabalPlatformFromSystem s
+
+-- | Parse a system tuple as understood by Nix and nixpkgs to a Cabal 'Platform'.
+--   System tuples are derived from autoconf's
+--   [target triplets](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.70/autoconf.html#Manual-Configuration),
+--   dropping the vendor part. They have the form @cpu-os@ where @os@ can either
+--   be a single component or of the form @kernel-system@ (system is an autoconf
+--   term here, not a Nix system).
+--
+--   Note that 'cabalPlatformFromSystem' expects to receive a /valid/ system
+--   tuple, i.e. it will accept all system tuples that have a sound format
+--   (with the caveat that it will accept n-tuples for @n >= 4@ even though
+--   they are technically invalid). This is done because the ambiguity of
+--   system tuples requires knowledge over its legal contents in order to check
+--   their validity properly. Since @lib.systems.elaborate@ from nixpkgs is the
+--   source of truth in this case, we want to avoid the need to continously
+--   update @distribution-nixpkgs@ to reflect its inner workings.
+--
+--   'cabalPlatformFromSystem' does, however, some conversions to alleviate some
+--   discrepancies between Cabal and nixpkgs. Parsing and rendering system tuples
+--   using 'cabalPlatformFromSystem' and rendering them via the 'Pretty'
+--   instance of 'NixpkgsPlatform' should, however, not change the system tuple
+--   for tuples accepted by nixpkgs. This has been tested for all known tuples
+--   (from @lib.platforms@ and @lib.systems.examples@) as of 2022-05-08.
+--   Please open an issue if any newly added ones are not recognized properly.
+cabalPlatformFromSystem :: String -> Maybe Platform
+cabalPlatformFromSystem s =
+  case break (== '-') s of
+    (arch, '-':os) ->
+      if null arch || null os
+      then Nothing
+      else Just $ Platform (parseArch arch) (parseOS os)
```

Comment:
```suggestion
    (arch, '-':os) -> if
      | null arch || null os = Nothing
      | otherwise -> Just $ Platform (parseArch arch) (parseOS os)
```

maybe nicer with `MultiWayIf`.